File: python/QEasyWidgets/QEasyWidgets/Components/GroupBox.py
from typing import Optional, overload
from PyEasyUtils import singledispatchmethod
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *

from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from .Widget import SizableWidget

##############################################################################################################################

class GroupBoxBase(QGroupBox, SizableWidget):
    """
    Base class for groupBox components
    """
    @singledispatchmethod
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)

        self.setCheckable(True)
        self.toggled.connect(lambda isChecked: self.expand() if isChecked else self.collapse())

        setFont(self, 15)

        StyleSheetBase.GroupBox.apply(self)

    @__init__.register
    def _(self, title: str, parent: Optional[QWidget] = None) -> None:
        self.__init__(parent)
        self.setTitle(title)

    def expand(self):
        setWidgetSizeAnimation(self, targetHeight = self.sizeHint().height()).start()

    def collapse(self):
        setWidgetSizeAnimation(self, targetHeight = self.fontMetrics().height() + 3).start()

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def setTransparent(self, transparent: bool) -> None:
        self.setProperty("isTransparent", transparent)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.GroupBox.deregistrate(self)

##############################################################################################################################

File: c++/QEasyWidgets/Components/GroupBox.cpp
#include "GroupBox.h"

#include <QFont>
#include <QPropertyAnimation>

#include "../Common/StyleSheet.h"
#include "../Common/QFunctions.h"


/**
 * GroupBoxBase implementation
 */

GroupBoxBase::GroupBoxBase(QWidget *parent)
    : SizableWidget<QGroupBox>(parent) {
    init();
}

GroupBoxBase::GroupBoxBase(const QString &title, QWidget *parent)
    : SizableWidget<QGroupBox>(title, parent) {
    init();
}

void GroupBoxBase::init() {
    setCheckable(true);
    connect(this, &QGroupBox::toggled, this, [this](bool checked) {
        if (checked) {
            expand();
        } else {
            collapse();
        }
    });

    QFont font = this->font();
    font.setPointSize(15);
    setFont(font);

    StyleSheetBase::apply(this, StyleSheetBase::GroupBox);
}

void GroupBoxBase::setBorderless(bool borderless) {
    setProperty("isBorderless", borderless);
}

void GroupBoxBase::setTransparent(bool transparent) {
    setProperty("isTransparent", transparent);
}

void GroupBoxBase::clearDefaultStyleSheet() {
    StyleSheetBase::deregistrate(this);
}

void GroupBoxBase::expand() {
    QParallelAnimationGroup *animation = setWidgetSizeAnimation(this, -1, sizeHint().height());
    if (animation) {
        animation->start(QPropertyAnimation::DeleteWhenStopped);
    }
}

void GroupBoxBase::collapse() {
    int titleHeight = fontMetrics().height() + 3;
    QParallelAnimationGroup *animation = setWidgetSizeAnimation(this, -1, titleHeight);
    if (animation) {
        animation->start(QPropertyAnimation::DeleteWhenStopped);
    }
}